Added Linux Commands (#13)
helicalco/docs: content/learn/using-linux.md

---
title: Using Linux
pcx-content-type: tutorial
weight: 1
meta:
  title: Using Linux
---

# Using Linux

## Useful Linux Commands

### cd

Allows you to change the current working directory.

```sh
$ cd /path/to/directory
```

### . & ..

`.` is the current directory. `..` is the parent directory.
For example, if you are in the directory `/home/user/Documents/my-project`, then `.` is `/home/user/Desktop/my-project` and `..` is `/home/user/Documents`. So to move backwards in the directory tree, you can use `cd ..`.

### ~

`~` is the home directory (`/home/user` by default).

### \*

Wildcard matches all files.

### ls

> Lists the contents of the specified directory.

Reviewer: ```suggestion
Lists the contents of the current directory.
```

```sh

> $ ls [directory]

Reviewer: ```suggestion
$ ls
```

```

> Can also just run `ls` alone and it will list the current directory.

Reviewer: ```suggestion
```

Running ls with -s flag will list files or directories with their sizes (in kilobytes).

```sh
$ ls -s [directory]
```

### cat

> Concatenate reads the contents of a file and prints it to the standard output.

Reviewer: ```suggestion
Concatenate reads the contents of a file and prints it to the terminal.
```

Lets keep it as simple as possible

```sh
$ cat [file]
```

### column

Column is super helpful when dealing with csv or any table formated files.

```sh

> $ column [file]

Reviewer: ```suggestion
$ column -t [file]
```

I would go as far as to suggest the -t flag as the default to use because that's the context its mostly used in

```

> Not much will happen differently from cat when using column with no flags. Two important flags are `-t` and `-s`. `-t` is used to create a table by determining the number of columns. `-s` defines the column delimiter. Using these together will format the output in a table.

Reviewer: ```suggestion
Additionally, the `-s` flag defines the column delimiter, allowing you to view CSV files in the same format.
When dealing with CSV files, make sure to provide the `-n` flag too which prevents adjacent delimiters being merged.
```

```sh

> $ column [file] -t -s ','

Reviewer: ```suggestion
$ column -t -n -s ',' [file]
```

```

> For example, you could have a genotype_report.txt file that looks likes this:

Reviewer: ```suggestion
For example, you could have a `genotype_report.txt` file that looks likes this:
```
I like wrapping filenames as inline code blocks!

```
ID CallRate #AA #AB #BB #NC %AA %AB %BB %NC
27813162 0.185837 141 192 142 2081 0.06 0.08 0.06 0.81
30282025 0.185837 115 240 120 2081 0.04 0.09 0.05 0.81
21928702 0.996870 596 951 1001 8 0.23 0.37 0.39 0.00
27942765 0.241393 142 292 183 1939 0.06 0.11 0.07 0.76
```

Which is a little hard to track columns. Using:

```sh

> $ column -t -s ' ' genotype_report.txt

Reviewer: ```suggestion
$ column -t genotype_report.txt
```

keep it as simple as possible

```

> Which is to say `-t` make a table, `-s ' '` seperate on spaces, and `genotype_report.txt` is the file to be read. We get:

Reviewer: ```suggestion
We get:
```
Don't really need this?

```
ID        CallRate  #AA  #AB  #BB   #NC   %AA   %AB   %BB   %NC
27813162  0.185837  141  192  142   2081  0.06  0.08  0.06  0.81
30282025  0.185837  115  240  120   2081  0.04  0.09  0.05  0.81
21928702  0.996870  596  951  1001  8     0.23  0.37  0.39  0.00
27942765  0.241393  142  292  183   1939  0.06  0.11  0.07  0.76
```

### echo

Echo is used to print text to the standard output.

```sh
$ echo [string]
```

> ### find

Reviewer: Could we either remove this or put it in an appendix? It's not a really important command for us

Finds files and directories.

```sh
$ find [directory to search from] [expression what to find] [what to find]
$ find Documents -name "*.md"
```

Above command will find all files with the extension `.md` in the `Documents` directory.

### grep

> Stands for - Globally search for a Regular Expression and Print it out.

Reviewer: ```suggestion
```
Don't need that!

> Preforms text searchs for a defined pattern.

Reviewer: ```suggestion
Finds all occurrences of text in a file.
```

```sh
$ grep [string/pattern] [files]
```

This will search for and return any lines of text that contain the pattern in given files.

> Helpful flags:

Reviewer: Great section! I like this

- `-i`: Case insensitive search.
- `-c`: Count the number of matches.
- `-n`: Print the line number of each match.

> ### history

Reviewer: Lets drop this too

Outputs your recent commands with numbers to identify them.

```sh
$ history
```

You can run a command again by using `!` and the number of the command `!5`.

### |

Pipe allows you to run multiple commands in a single command. It pipes the standard output of one command into the standard input of the next command.

```sh

> $ hisory | grep cd

Reviewer: ```suggestion
$ cat pedigree.txt | column -t
```

```

> The above command would list the history which would pipe into grep. Grep would then search for `cd` and output every command you've made with cd.

Reviewer: ```suggestion
The above example would print the file out and pipe the contents into `column`, which would then format the file.
```

### head

Prints the first n lines of a file/input.

```sh
$ head [file]
```

Can specifiy number of lines using `-n` flag.

### man

Lookup manual for a given command.

```sh
$ man [command]
```

### mkdir

> Creates directories.

Reviewer: ```suggestion
Creates a new directory.
```

```sh

> $ mkdir [directories]

Reviewer: ```suggestion
$ mkdir my-directory
```

```

> `-p` flag will create parent as necessary ie. can create 3 directories one inside the next.

Reviewer: ```suggestion
```

> ```sh
> $ mkdir -p dir1/dir2/dir3
> ```

Reviewer: ```suggestion
```
